pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +257,297 @@ def simulate(
+
+
+def find_not_measurement_values(
+    circuit: qml.tape.QuantumScript,
+):
+    """Returns the indices of measurements with a trivial measurement value.
+
+    Args:
+        circuit (QuantumTape): A QuantumScript
+
+    Returns:
+        List[int]: Indices of measurements with a trivial measurement value.
+    """
+    return [i for i, m in enumerate(circuit.measurements) if not has_measurement_values(m)]
+
+
+def parse_native_mid_circuit_measurements(
+    circuit: qml.tape.QuantumScript, all_shot_meas, mcm_shot_meas
+):
+    """Combines, gathers and normalizes the results of native mid-circuit measurement runs.
+
+    Args:
+        circuit (QuantumTape): A one-shot (auxiliary) QuantumScript
+        all_shot_meas (Sequence[Any]): List of accumulated measurement results
+        mcm_shot_meas (Sequence[dict]): List of dictionaries containing the mid-circuit measurement results of each shot
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    idx_one_shot = find_not_measurement_values(circuit)
+    normalized_meas = [None] * len(circuit.measurements)
+    for i, m in zip(idx_one_shot, all_shot_meas):
+        normalized_meas[i] = gather_non_mcm(circuit.measurements[i], m, mcm_shot_meas)
+
+    idx_sample = find_measurement_values(circuit)
+    counter = Counter()
+    if any(isinstance(m, (CountsMP, ProbabilityMP)) for m in circuit.measurements):
+        for d in mcm_shot_meas:
+            counter.update(d)
+    for i in idx_sample:
+        normalized_meas[i] = gather_mcm(
+            circuit.measurements[i], circuit.measurements[i].mv, mcm_shot_meas, counter
+        )
+    return tuple(normalized_meas) if len(normalized_meas) > 1 else normalized_meas[0]
+
+
+def gather_non_mcm(circuit_measurement, measurement, samples):
+    """Combines, gathers and normalizes several measurements with trivial measurement values.
+
+    Args:
+        circuit_measurement (MeasurementProcess): measurement
+        measurement (TensorLike): measurement results
+        samples (List[dict]): Mid-circuit measurement samples
+
+    Returns:
+        TensorLike: The combined measurement outcome
+    """
+    if isinstance(circuit_measurement, (ExpectationMP, ProbabilityMP)):
+        new_meas = measurement / len(samples)
+    elif isinstance(circuit_measurement, SampleMP):
+        new_meas = np.concatenate(tuple(s.ravel() for s in measurement))
+    elif isinstance(circuit_measurement, CountsMP):
+        new_meas = dict(sorted(measurement.items()))
+    elif isinstance(circuit_measurement, VarianceMP):
+        new_meas = qml.math.var(np.concatenate(tuple(s.ravel() for s in measurement)))
+    else:
+        raise ValueError(
+            f"Native mid-circuit measurement mode does not support {circuit_measurement.__class__.__name__} measurements."
+        )
+    return new_meas
+
+
+def gather_mcm(measurement, mv, samples, counter):
```

Comment:
What would you think about turning this into a `functools.singledispatch` function?

Reply:
For example:

```
@singledispatch
def gather_mcm(measurement, mv, samples, counter):
    """Combines, gathers and normalizes several measurements with non-trivial measurement values.

    Args:
        measurement (MeasurementProcess): measurement
        mv (MeasurementValue): measurement value
        samples (List[dict]): Mid-circuit measurement samples
        counter (Counter): Measurement value counts

    Returns:
        TensorLike: The combined measurement outcome
    """
    raise ValueError(
        f"Native mid-circuit measurement mode does not support {measurement.__class__.__name__} measurements."
    )

@gather_mcm.register
def _(measurement: ExpectationMP, mv, samples, counter):
    sha = mv.measurements[0].hash
    return np.mean(np.array([dct[sha] for dct in samples]))

@gather_mcm.register
def _(measurement: ProbabilityMP, mv, samples, counter):
    sha = mv.measurements[0].hash
    p1 = counter[sha] / float(len(samples))
    return np.array([1 - p1, p1])

@gather_mcm.register
def _(measurement: CountsMP, mv, samples, counter):
    sha = mv.measurements[0].hash
    new_samples = {}
    if len(samples) - counter[sha] > 0:
        new_samples[0] = len(samples) - counter[sha]
    if counter[sha] > 0:
        new_samples[1] = counter[sha]
    return new_samples
```

Reply:
Do you mean using `singledispatch` to merge the loops and call `gather_non_mcm`, `gather_mcm` based on input?

Reply:
Nice idea. `gather_mcm` will change quite a bit in the follow-up PR, but I'll reuse the pattern. Thanks.